src/polyswarmd/artifacts/ipfs.py

```diff
@@ -0,0 +1,256 @@
+
+    return wrapper
+
+
+class IpfsServiceClient(AbstractArtifactServiceClient):
+    """
+    Artifact Service Client for IPFS.
+
+    Uses MFS for adding to directories, since limits on IPFS API requests prevent 256 file requests.
+    """
+    def __init__(self, base_uri):
+        self.base_uri = base_uri
+        reachable_endpoint = "{}{}".format(self.base_uri, '/api/v0/bootstrap')
+        super().__init__('IPFS', reachable_endpoint)
+
+    @staticmethod
+    def check_ls(artifacts, index, max_size=None):
+        if not artifacts:
+            raise ArtifactServiceException(404, 'Could not locate IPFS resource')
+
+        if index < 0 or index > 256 or index >= len(artifacts):
+            raise ArtifactServiceException(404, 'Could not locate artifact ID')
+
+        _, artifact, size = artifacts[index]
+        if max_size and size > max_size:
+            raise ArtifactServiceException(400, 'Artifact size greater than maximum allowed')
+
+        return artifacts[index]
+
+    @staticmethod
+    def check_redis(uri, redis):
+        try:
+            result = redis.get('polyswarmd:{0}'.format(uri))
+            if result:
+                return result
+        except RuntimeError:
+            # happens if redis is not configured and websocket poll calls this
+            pass
+
+    def add_artifacts(self, artifacts, session):
+        directory = self._mfs_mkdir(session)
+        for artifact in artifacts:
+            ipfs_uri = self._add(artifact, session)
+            filename = artifact[1][0]
+            self._mfs_copy(directory, filename, ipfs_uri, session)
+
+        stat = self._mfs_stat(directory, session)
+        return stat.get('Hash', '')
+
+    def add_artifact(self, artifact, session, redis=None):
+        ipfs_uri = self._add(artifact, session)
+        if redis:
+            redis.set('polyswarmd:{0}'.format(ipfs_uri), artifact[1][1], ex=300)
+
+        return ipfs_uri
+
+    # noinspection PyBroadException
+    def check_uri(self, uri):
+        # TODO: Further multihash validation
+        try:
+            return len(uri) < 100 and base58.b58decode(uri)
+        except Exception:
+            return False
+
+    # noinspection PyBroadException
+    def details(self, uri, index, session):
+        if not self.check_uri(uri):
+            raise ArtifactServiceException(400, 'Invalid IPFS Hash')
+
+        artifacts = self.ls(uri, session)
+        name, artifact, _ = IpfsServiceClient.check_ls(artifacts, index)
+
+        stat = self._stat(artifact, session)
+
+        # Convert stats to snake_case
+        stats = {
+            re.sub(r'(.)([A-Z][a-z]+)', r'\1_\2', k).lower(): v
+            for k, v in stat.items()
+        }
+        stats['name'] = name
+
+        return stats
+
+    def get_artifact(self, uri, session, index=None, max_size=None, redis=None):
+        if not self.check_uri(uri):
+            raise ArtifactServiceException(400, 'Invalid IPFS Hash')
```

Comment:
I have the impression that we are mixing responsibilities here. This method should not be concerned with web server status codes, it should only raise the exception with a message. If this was called from a webserver view (which might not always be the case), this exception should be caught there and the proper status code for this particular exception should be rendered. This helps making things more decoupled and having views worried about status codes and the rest of the code worried about exceptions. It is really good to use exceptions in general rather than generating responses in the middle of the code. :+1:

Reply:
Agreed.

I do want to forward the status_code from an artifact service along in some cases, but that can be the exception, rather than the rule. And I can just let the HTTPError do that.